cranelift/codegen/src/isa/s390x/lower.rs

```diff
@@ -2734,37 +2734,61 @@ fn lower_insn_to_regs<C: LowerCtx<I = Inst>>(
                 ctx.emit(Inst::AtomicCas64 { rd, rn, mem });
             }
         }
-        Opcode::AtomicLoad => {
+        Opcode::AtomicLoad
+        | Opcode::AtomicUload8
+        | Opcode::AtomicUload16
+        | Opcode::AtomicUload32 => {
             let flags = ctx.memflags(insn).unwrap();
             let endianness = flags.endianness(Endianness::Big);
             let ty = ty.unwrap();
-            assert!(is_valid_atomic_transaction_ty(ty));
+            let access_ty = match op {
+                Opcode::AtomicLoad => ty,
+                Opcode::AtomicUload8 => types::I8,
+                Opcode::AtomicUload16 => types::I16,
+                Opcode::AtomicUload32 => types::I32,
+                _ => unreachable!(),
+            };
+            assert!(is_valid_atomic_transaction_ty(access_ty));
 
             let mem = lower_address(ctx, &inputs[..], 0, flags);
             let rd = get_output_reg(ctx, outputs[0]).only_reg().unwrap();
 
             if endianness == Endianness::Big {
-                ctx.emit(match ty_bits(ty) {
-                    8 => Inst::Load32ZExt8 { rd, mem },
-                    16 => Inst::Load32ZExt16 { rd, mem },
-                    32 => Inst::Load32 { rd, mem },
-                    64 => Inst::Load64 { rd, mem },
+                ctx.emit(match (ty_bits(access_ty), ty_bits(ty)) {
+                    (8, 32) => Inst::Load32ZExt8 { rd, mem },
+                    (8, 64) => Inst::Load64ZExt8 { rd, mem },
+                    (16, 32) => Inst::Load32ZExt16 { rd, mem },
+                    (16, 64) => Inst::Load64ZExt16 { rd, mem },
+                    (32, 32) => Inst::Load32 { rd, mem },
+                    (32, 64) => Inst::Load64ZExt32 { rd, mem },
+                    (64, 64) => Inst::Load64 { rd, mem },
                     _ => panic!("Unsupported size in load"),
                 });
             } else {
-                ctx.emit(match ty_bits(ty) {
-                    8 => Inst::Load32ZExt8 { rd, mem },
-                    16 => Inst::LoadRev16 { rd, mem },
-                    32 => Inst::LoadRev32 { rd, mem },
-                    64 => Inst::LoadRev64 { rd, mem },
+                ctx.emit(match (ty_bits(access_ty), ty_bits(ty)) {
+                    (8, 32) => Inst::Load32ZExt8 { rd, mem },
+                    (8, 64) => Inst::Load64ZExt8 { rd, mem },
+                    (16, 32) => Inst::LoadRev16 { rd, mem },
```

Comment:
This is wrong, because the lrvh instruction does not perform any extension.   To support this, you'd have to do the same that is currently done for a Uload16 and emit an explicit Extend after the load-reverse.

Also, I think you need to support the same set of operations for the little-endian case as for the big-endian case, which means the 16->64 and 32->64 extension cases also should be added.  They need to be implemented via the same explicit Extend after the load.

Reply:
Okay, thanks.